crowdnotifier-sdk/src/main/java/org/crowdnotifier/android/sdk/utils/CryptoUtils.java

```diff
@@ -43,15 +56,20 @@ public static synchronized CryptoUtils getInstance() {
 	public EncryptedVenueVisit getEncryptedVenueVisit(long arrivalTime, long departureTime, VenueInfo venueInfo) {
 
 		G2 masterPublicKey = new G2();
-		masterPublicKey.deserialize(venueInfo.getMasterPublicKey());
+		masterPublicKey.deserialize(venueInfo.getPublicKey());
 
 		//scan
 		ArrayList<IBECiphertext> ibeCiphertextsEntries = new ArrayList<>();
 
 		ArrayList<Integer> hourCounters = getAffectedHours(arrivalTime, departureTime);
 		for (Integer hour : hourCounters) {
 
-			byte[] identity = generateIdentity(hour, venueInfo);
+			byte[] identity;
+			if (venueInfo.getQrCodePayload() == null) {
+				identity = generateIdentityV2(hour, venueInfo);
+			} else {
+				identity = generateIdentityV3(venueInfo.getQrCodePayload(), hour * 3600L);
```

Comment:
This might be worth a little comment of the form:

```
// getAffectedhours() generates hours since UNIX epoch. As the new format requires these to be in seconds since
// UNIX epoch, we multiply by 60 minutes * 60 seconds = 3600.
```